src/matrix_observer.py

```diff
@@ -24,19 +31,37 @@ def __init__(self, charm: CharmBase, relation_name: str):
         super().__init__(charm, "matrix-observer")
         self._charm = charm
         self.relation_name = relation_name
+        self.matrix = MatrixAuthRequires(
+            self._charm,
+            relation_name=relation_name,
+        )
+        self.framework.observe(
+            self.matrix.on.matrix_auth_request_processed,
+            self._on_matrix_auth_request_processed,
+        )
 
-    def _get_relation_data(self) -> typing.Optional[DatasourceMatrix]:
-        """Get matrix data from relation.
+    def _on_matrix_auth_request_processed(self, _: Object) -> None:
+        """Handle the matrix auth request processed event."""
+        logger.info("Matrix auth request processed")
+        self._charm.reconcile()  # type: ignore
+
+    def get_matrix(self) -> typing.Optional[MatrixAuthProviderData]:
+        """Return a Matrix authentication datasource model.
 
         Returns:
-            Dict: Information needed for setting environment variables.
+            MatrixAuthProviderData: The datasource model.
         """
-        return DatasourceMatrix(host="localhost")
+        return self.matrix.get_remote_relation_data()
 
-    def reconcile(self) -> typing.Optional[DatasourceMatrix]:
-        """Reconcile the database relation.
+    def set_irc_registration(self, content: str) -> None:
+        """Set the IRC registration details.
 
-        Returns:
-            Dict: Information needed for setting environment variables.
+        Args:
+            content: The registration content.
         """
-        return self._get_relation_data()
+        registration = typing.cast(SecretStr, content)
+        irc_data = MatrixAuthRequirerData(registration=registration)
+        relation = self.model.get_relation(self.matrix.relation_name)
+        if relation:
+            irc_data.set_registration_id(model=self.model, relation=relation)
+            self.matrix.update_relation_data(relation=relation, matrix_auth_requirer_data=irc_data)
```

Comment:
I'm not following this part. Why update_relation_data?

Reply:
It explicitly updates the relation data for consistency after storing a secret and triggers a relation changed. I think that's the way to go, wdyt?